Updated T4 templates

gabog opened the pull request:
Hi, I updated the T4 templates with a new version that dynamically determines the location of the json file during code gen so we don't need to alter the template for each new resources we create.

Let me know if you have any issues or questions

darrenj commented:
@lzc850612 made some changes yesterday which are in master around adding better exception handling to handle files JSON files with errors - you'll likely have to do a merge?

https://github.com/Microsoft/AI/pull/168/files

gabog commented:
Hi Ted, I didn't know about the include directive, let me play with it a bit and I will update the PR if I can figure out the best place to put the shared file.

lzc850612 commented:
@gabog sure thing. I think we can put it in microsoft.bot.solutions project, under Resources. Then it'll be able to be referenced by all skills.

gabog commented:
Hi @lzc850612 and @darrenj . I updated the code to use a shared t4 template approach you suggested.

In addition to that, I also:
	• Extracted the logic to read and create responses out of the generated code an put it in a separate ResponseManager class. 
	This allowed me to create unit tests and address some of the common issues we have more explicitly.
	It will also allow devs to use a different resource manager in the future if they want without having to regenerate the code.
	• I also change the logic so we don't assume English as the fallback language. So I renamed all the .en.json files we have as .json, these will be the default and it doesn't need to be english, other people can assume the default is spanish and add en.json later. (Resx style).

Please review and let me know what you think.

Gabo

gabog commented:
btw, I added 8 more tests but they don't show in the build results, if you approve the PR, you may need to update the VA-PRValidation to include the new tests for Microsoft.Bot.Solutions.Tests.

Gabo

lzc850612 commented:
> btw, I added 8 more tests but they don't show in the build results, if you approve the PR, you may need to update the VA-PRValidation to include the new tests for Microsoft.Bot.Solutions.Tests.
> 
> Gabo

Thanks Gabo. I think it's probably because of the test filter i put in the build definition. I've just made the change so it'll run the tests after your change is merged. Or if you need to push a new iteration, it'll trigger another build so it'll pick up my change the run your tests.

lzc850612 commented on solutions/Virtual-Assistant/src/csharp/skills/calendarskill/Dialogs/Shared/Resources/CalendarSharedResponses.cs:
```diff
@@ -1,124 +1,44 @@
-                        _jsonResponses = null;
-                        resourceLoadingException = ex;
-                        break;
-                    }
```
These try catch are added by me for the shared responses in four skills. I didn't put it in all response.cs files but i should. Can you add this into the common .t4 file? it's needed for better error handling.

lzc850612 replied:
actually let me add them. i'll have a better idea how to test it.

gabog commented:
OK. Thanks. I saw those but couldn’t come up with a test that required that try/catch and I didn't port it after refactoring everything to Response Manager.

Sent from mobile
________________________________
From: Ted <notifications@github.com>
Sent: Friday, October 26, 2018 7:11:13 PM
To: Microsoft/AI
Cc: Gabo Gilabert; Mention
Subject: Re: [Microsoft/AI] Updated T4 templates (#173)


@lzc850612 approved this pull request.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2FAI%2Fpull%2F173%23pullrequestreview-169026779&data=02%7C01%7CGabo.Gilabert%40microsoft.com%7C37ae3fe33c274614416008d63b9852f6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636761922745225797&sdata=RkCMRgXy5M9pF40%2BOBF6wYki9uxWPTsCfGI2%2FjrJ2aI%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FALsl8qdw8H4jPRzx6MicgwJog3NTmrJnks5uo5aRgaJpZM4X3tkx&data=02%7C01%7CGabo.Gilabert%40microsoft.com%7C37ae3fe33c274614416008d63b9852f6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636761922745225797&sdata=reIYW22mVSymCz8w%2FJWUCQ77tBfUbfRU4CYpIbit1IQ%3D&reserved=0>.
